lib/Transforms/LLZKPCLLoweringPass.cpp

```diff
@@ -0,0 +1,427 @@
+      }
+      return false;
+    };
+    auto isConstZero = [&](mlir::Value v) {
+      if (auto ap = getConstAPInt(v)) {
+        return ap->isZero();
+      }
+      return false;
+    };
+
+    auto emitEqAssert = [&](mlir::Value l, mlir::Value r) {
+      auto eq = b.create<pcl::CmpEqOp>(loc, l, r);
+      b.create<pcl::AssertOp>(loc, eq.getRes());
+    };
+
+    auto emitAssertTrue = [&](mlir::Value pred) { b.create<pcl::AssertOp>(loc, pred); };
+
+    auto emitAssertFalse = [&](mlir::Value pred) {
+      auto neg = b.create<pcl::NotOp>(loc, pred);
+      b.create<pcl::AssertOp>(loc, neg.getRes());
+    };
+
+    // Optimized handling of boolean patterns
+    if (isBool(lhsVal) && isConstOne(rhsVal)) {
+      // bool == 1 → assert(bool)
+      emitAssertTrue(lhsVal);
+      return success();
+    }
+    if (isBool(rhsVal) && isConstOne(lhsVal)) {
+      // 1 == bool → assert(bool)
+      emitAssertTrue(rhsVal);
+      return success();
+    }
+    if (isBool(lhsVal) && isConstZero(rhsVal)) {
+      // bool == 0 → assert(!bool)
+      emitAssertFalse(lhsVal);
+      return success();
+    }
+    if (isBool(rhsVal) && isConstZero(lhsVal)) {
+      // 0 == bool → assert(!bool)
+      emitAssertFalse(rhsVal);
+      return success();
+    }
+
+    // Fallback to assert(lhs == rhs)
+    emitEqAssert(lhsVal, rhsVal);
+    return success();
+  }
+
+  /// Lower the constraint ops to PCL ops
+  LogicalResult lowerStructToPCLBody(
+      StructDefOp structDef, func::FuncOp dstFunc, llvm::DenseMap<Value, Value> &llzkToPcl
+  ) {
+    OpBuilder b(dstFunc.getBody());
+    // Map field name to PCL vars; public fields are outputs, privates are intermediates
+    llvm::DenseMap<StringRef, Value> field2pclvar;
+    llvm::SmallVector<Value> outVars;
+
+    auto srcFunc = structDef.getConstrainFuncOp();
+    auto srcArgs = srcFunc.getArguments().drop_front();
+    auto dstArgs = dstFunc.getArguments();
+    if (dstArgs.size() != srcArgs.size()) {
+      return srcFunc.emitError("arg count mismatch after dropping self");
+    }
+
+    // 1-1 mapping of args from constraint args to PCL args
+    for (auto [src, dst] : llvm::zip(srcArgs, dstArgs)) {
+      llzkToPcl.try_emplace(src, dst);
+    }
+    for (auto fieldDef : structDef.getFieldDefs()) {
+      // Create a PCL var for each struct field. Public fields are outputs in PCL
+      auto pclVar =
+          b.create<pcl::VarOp>(fieldDef.getLoc(), fieldDef.getName(), fieldDef.hasPublicAttr());
+      field2pclvar.insert({fieldDef.getName(), pclVar});
+      if (fieldDef.hasPublicAttr()) {
+        outVars.push_back(pclVar);
+      }
+    }
+    if (!srcFunc.getBody().hasOneBlock()) {
+      return srcFunc.emitError(
+          "llzk-to-pcl translation assumes the constrain function body has 1 block"
+      );
+    }
+    Block &srcEntry = srcFunc.getBody().front();
+    // Translate each op. Almost 1-1 and currently only support Felt ops.
+    // TODO: support calls.
+    for (Operation &op : srcEntry) {
```

Comment:
Things like loops and if-then-elses are not handled here either, tho picus supports the latter. Do we plan on supporting those? If not I'd suggest mentioning that this pass requires running the inliner prior to this